### ISSUE TITLE
Can delimitMate expand <CR> feature be achieved via ultisnips?

### ISSUE THREAD

**Issue by purpleP** [SirVer/ultisnips#782]
When I first learned about autotriggers in ultisnips I've immediately  thought: "Hey, this reminds me of something". And the thing it reminded me about was delimitMate plugin (there are other plugins that provides the same functionality).
You know when I was using IDE it was nice that if I would type
`(` it would be automatically converted to `(|)` where `|` is cursor position. And if I would type enter after that it would be converted to

```
(
    |
)
```

etc.

Also feature that I kind of miss is when editing xml this `<tag>` would be converted to this `<tag>|</tag>` right after I would type last `>` and again if I would type enter again it would expand it to

```
<tag>
    |
</tag>
```

I've read about generating tabstops dynamically and managed to create a snippet that would be triggered by regexp "[([{]" and in after_jump action I would get the character that were typed from `char = snip.buffer[snip.line][snip.column]` and returning something like this `char + '$1' + matching_char(char)` which would give me the automatching open parens, brackets etc. But I couldn't get expanding of enter to work. My attempt was to add

```
snippet "\[\n\]"
[
    $1
]
```

But this isn't working probably because ultisnips checks only one line of input for regex match.

So to summarize functionality provided by delimitMate and IDE's seems to be very close to Ultisnips. So if it isn't possible with Ultisnips now maybe it should be?

Conceptually it seems to be easy. Keep track of what user is typing and try to match it with all patterns in user defined regex snippets in priority order. And you probably already doing that also.

**Comment by seletskiy** [comment]
@purpleP: after `9a670c5` it should be more or less possible (issue was that vim is not catching `Enter` as text modification character, e.g. leaving `v:char` empty).

Regexp-based snippets will match only current line indeed, but to match other lines there is `context` feature (`:h UltiSnips-context-snippets`).

After mentioned commit it's possible to autoindent `[\n]` with following snipeet (ugly!):
```
context "snip.buffer[snip.line-1].endswith('[') and snip.buffer[snip.line].strip().startswith(']')"
snippet "" "[\n]" brA
    $0

endsnippet
```

Note, that snippet will try to trigger on every change due empty regexp condition, so it can lead to slowing down when entering text.

Also note, that UltiSnips will not be good at pair-completing task, it doesn't have specific features that pair-completing plugins has.

**Comment by purpleP** [comment]
@seletskiy Yeah, I've already dumped that idea. In fact I've read docs of delimitMate more closely and I've been able to set it to behave the way I wanted, so it's no more giving me frustration.
`let b:delimitMate_jump_expansion = 0`

**Comment by seletskiy** [comment]
@purpleP: soooo, I'm closing issue?
